app/waveplot/waveplot.F90

```diff
@@ -237,135 +236,142 @@ program waveplot
+            & buffer, fileName, comments=comments, repeatBox=wp%opt%repeatBox)
+        write(*, "(A)") "File '" // trim(fileName) // "' written"
       end if
 
       if (wp%opt%tPlotChrgDiff) then
-        buffer = buffer - (sumChrg / sumAtomicChrg) * atomicChrg(:,:,:,1)
-        write (comments(2), 9995) wp%input%identity, iSpin, iKPoint, iLevel
+        buffer(:,:,:) = buffer - (sumChrg / sumAtomicChrg) * atomicChrg(:,:,:,1)
+        write(comments(2), 9995) wp%input%identity, iSpin, iKPoint, iLevel
 9995    format('Calc-Id:',I11,', Spin:',I2,', K-Point:',I6,', State:',I6, ', abs2diff')
         fileName = "wp-" // i2c(iSpin) // "-" // i2c(iKPoint) // "-" // i2c(iLevel) //&
             & "-abs2diff.cube"
         call writeCubeFile(wp%input%geo, wp%aNr%atomicNumbers, wp%loc%gridVec, wp%opt%gridOrigin,&
-            & buffer, fileName, comments, wp%opt%repeatBox)
-        write(stdout, "(A)") "File '" // trim(fileName) // "' written"
+            & buffer, fileName, comments=comments, repeatBox=wp%opt%repeatBox)
+        write(*, "(A)") "File '" // trim(fileName) // "' written"
       end if
 
       if (wp%opt%tPlotReal) then
         if (wp%input%tRealHam) then
-          buffer = gridValReal
+          buffer(:,:,:) = gridValReal
         else
-          buffer = real(gridValCmpl, dp)
+          buffer(:,:,:) = real(gridValCmpl, dp)
         end if
-        write (comments(2), 9991) wp%input%identity, iSpin, iKPoint, iLevel
+        write(comments(2), 9991) wp%input%identity, iSpin, iKPoint, iLevel
 9991    format('Calc-Id:',I11,', Spin:',I2,', K-Point:',I6,', State:',I6, ', real')
         fileName = "wp-" // i2c(iSpin) // "-" // i2c(iKPoint) // "-" // i2c(iLevel) // "-real.cube"
         call writeCubeFile(wp%input%geo, wp%aNr%atomicNumbers, wp%loc%gridVec, wp%opt%gridOrigin,&
-            & buffer, fileName, comments, wp%opt%repeatBox)
-        write(stdout, "(A)") "File '" // trim(fileName) // "' written"
+            & buffer, fileName, comments=comments, repeatBox=wp%opt%repeatBox)
+        write(*, "(A)") "File '" // trim(fileName) // "' written"
       end if
 
       if (wp%opt%tPlotImag) then
-        buffer = aimag(gridValCmpl)
-        write (comments(2), 9992) wp%input%identity, iSpin, iKPoint, iLevel
+        buffer(:,:,:) = aimag(gridValCmpl)
+        write(comments(2), 9992) wp%input%identity, iSpin, iKPoint, iLevel
 9992    format('Calc-Id:',I11,', Spin:',I2,', K-Point:',I6,', State:',I6, ', imag')
         fileName = "wp-" // i2c(iSpin) // "-" // i2c(iKPoint) // "-" // i2c(iLevel) // "-imag.cube"
         call writeCubeFile(wp%input%geo, wp%aNr%atomicNumbers, wp%loc%gridVec, wp%opt%gridOrigin,&
-            & buffer, fileName, comments, wp%opt%repeatBox)
-        write(stdout, "(A)") "File '" // trim(fileName) // "' written"
+            & buffer, fileName, comments=comments, repeatBox=wp%opt%repeatBox)
+        write(*, "(A)") "File '" // trim(fileName) // "' written"
```

Comment:
Why not stdout? Also, should the lead proc be writing this only?

Reply:
Because each MPI rank owns different entries of the composite index and there should be a message printed for each index processed. This also means that all processes write their own stuff (individual grids are not distributed at the moment), which collides with the logic in `file.F90` (see the changes in `writeCubeFile()`, that we have to discuss).

Reply:
https://github.com/dftbplus/dftbplus/blob/6002d2ec8ebcfc0ce2cba2db8b8e09208d10dd49/src/dftbp/common/file.F90#L285-L299

This code currently blocks I/O with ranks other than the lead process, which collides with what I'm trying to achieve in Waveplot. What I'm not aware of is if my current strategy is actually dangerous. To me it is the logical thing to do, should the filesystem be able to handle this in a save manner and should we be fine with not distributing individual grids. My opinion on this is that parallelizing over the composite index only is elegant for both small (many k-points) and large (many states) systems. Should we ever reach the point where this strategy (in combination with `NrOfCachedGrids=1`) is too memory consuming, we are probably screwed anyway.

Reply:
Could you collect the messages and either the error state (MPI_LAND ?) or the proc numbers in error state on the global lead processor and print there before a global abort and stop?

Reply:
But collecting the messages on the lead only delays printing them, the order will be just as non-deterministic. Is this just about restricting any stdout printing to the lead process or am I missing the point here?

Regarding the error state, you suggest that the `writeCubeFile()` routine passes back the `iostat` integer and at the end of the `lpStates` loop we collect all such error states via an `MPI_LAND` operation and possibly abort? I'm just trying to understand. What do you mean by 'the proc numbers in error state'?